src/main/java/rx/internal/operators/OperatorMerge.java

```diff
@@ -71,715 +72,878 @@
+                    if (o == null) {
+                        o = new Object[4];
+                        fastArray = o;
+                    } else
+                    if (s == o.length) {
+                        Object[] o2 = new Object[s + (s >> 2)];
+                        System.arraycopy(o, 0, o2, 0, s);
+                        fastArray = o2;
+                        o = o2;
                     }
+                    o[s] = value;
+                    fastSize = s + 1;
+                    missed = true;
+                    return;
                 }
-                exceptions.add(e);
-                boolean sendOnComplete = false;
-                synchronized (this) {
-                    if (!parent) {
-                        wip--;
+                emitting = true;
+            }
+            boolean skipFinal = false;
+            try {
+                child.onNext(value);
+                for (int k = Integer.MAX_VALUE; k >= 0; k--) {
+                    Object[] a;
+                    int s;
+                    synchronized (this) {
+                        if (!missed) {
+                            emitting = false;
+                            skipFinal = true;
+                            return;
+                        }
+                        a = fastArray;
+                        s = fastSize;
+                        fastArray = null;
+                        fastSize = 0;
+                        missed = false;
                     }
-                    if ((wip == 0 && completed) || (wip < 0)) {
-                        sendOnComplete = true;
+                    for (int i = 0; i < s; i++) {
+                        Object v = a[i];
+                        if (v == REPORT_ERROR) {
+                            reportError();
+                            skipFinal = true;
+                            return;
+                        } else
+                        if (v == REPORT_COMPLETED) {
+                            child.onCompleted();
+                            skipFinal = true;
+                            return;
+                        }
+                        @SuppressWarnings("unchecked")
+                        T v2 = (T)v;
+                        child.onNext(v2);
                     }
                 }
-                if (sendOnComplete) {
-                    drainAndComplete();
+            } finally {
+                if (!skipFinal) {
+                    synchronized (this) {
+                        emitting = false;
+                    }
                 }
-            } else {
-                actual.onError(e);
             }
         }
-
-        @Override
-        public void onCompleted() {
-            boolean c = false;
+        void fastEmitError() {
             synchronized (this) {
-                completed = true;
-                if (wip == 0) {
-                    c = true;
+                if (emitting) {
+                    Object[] o = new Object[1];
```

Comment:
Since when do we report errors eagerly? The only place I'm aware of that decision ever being made is in `observeOn` due to the concurrent nature of thread-hopping. 

We didn't do that with merge before, so why now?